### PR TITLE
Performance improvements

### PR DESCRIPTION
**Pull request by riha112**
Performance improvements:

Testing:
* Used Lighthouse on Chrome

In this PR:
* Bundle optimization - splitted chunks into smaller ones by pages, tabs, widgets, sections & nameless chunk joining.
* Moved constant values that caused chunk enlargement from .container.js file into .config.js file
* Implementation of new utility ```withReducers``` for dynamic reducer usage.
* Removed specific-case reducers from static reducer batch and moved them accordingly to they use-case to be loaded dynamically
* Code improvements:
    * Reduced re-rendering count for components via ```shouldComponentUpdate```
    * Removed inline function from JSX properties.
    * Optimized conditional rendering.
    * Moved external font to be loaded async
* Checked whole code-base:
    * for possible implementations of React.memo, useMemo & css animations.
    * for debouncing & throttling implementations

### CODE REVIEW

**Comment by zans-laksa** [comment]
I may be wrong, but when you do something like this
```javascript
shouldComponentUpdate(nextProps) {
    const { items } = this.props;
    const { items: nextItems } = nextProps;

   return items !== nextItems;
}
```
Don't you do the same thing the PureComponent does - the shallow comparison?
But when PureComponent compares all properties, you compare only one.

And when we do something like this
```javascript
return (
    <SomeComponent someProp={ { ...someData, ...someOtherData } } />
);
```
Then `items` and `nextItems` will never be equal when compare, since they have different references.

File: packages/scandipwa/src/component/CheckoutOrderSummary/CheckoutOrderSummary.component.js [hunk 1]
```diff
@@ -102,16 +102,11 @@ export class CheckoutOrderSummary extends PureComponent {
             return null;
         }
 
-        if (!coupon_code) {
-            return this.renderPriceLine(
-                -Math.abs(discount_amount),
-                __('Discount %s:', '')
-            );
-        }
+        const label = coupon_code ? __('Discount/Coupon %s:', coupon_code.toUpperCase()) : __('Discount %s:', '');
```
Comment (alfredsgenkins) [inline packages/scandipwa/src/component/CheckoutOrderSummary/CheckoutOrderSummary.component.js:105]: Very strange line, I do not understand what it does: `__('Discount %s:', '')`, hmm!

File: packages/scandipwa/src/component/RecentlyViewedWidget/RecentlyViewedWidget.container.js [hunk 2]
```diff
@@ -103,4 +105,6 @@ export class RecentlyViewedWidgetContainer extends PureComponent {
     }
 }
 
-export default connect(mapStateToProps, mapDispatchToProps)(RecentlyViewedWidgetContainer);
+export default withReducers({
```
Comment (alfredsgenkins) [inline packages/scandipwa/src/component/RecentlyViewedWidget/RecentlyViewedWidget.container.js:108]: WOW, you are the GOD! With reducer is awesome feature!!!

**Review by alfredsgenkins — APPROVED** [review]
Wow, `withReducer` is amazing. Feels like our bundle is nearing 100kb right now.

What effect did the should update introduction had?

**Comment by riha112** [comment]
@zans-laksa @alfredsgenkins 
Regarding shouldComponentUpdate vs PureComponent

It is correct that PureComponent will perform shallow comparison, however it will be done for all properties, meaning that each time one property changes value the render() method will be called (simple to see with console.log), sometimes this can cause to call render() method when not needed (for example Header is called up to 8 times on load) in my implementations my goal was to reduce exactly that, by following one of the following methods/rules
* Implement shouldComponentUpdate for those components that have one or two dependent properties
* Implement shouldComponentUpdate if is known last loaded property (example would be that we are passing functions + product to component, and we know that functions will be passed first as their reference is defined, but product may still be loaded, so on first pass product will be undefined and on second time defined), thus we can reduce render() call from 2 to 1.

Another minor improvement would be that one property comparison costs less than reacts shallow comparison (but that's a micro improvement)

Regarding example where we are passing deconstructed values to component and comparing them with previous or next values (depends on how you look at it) it won't always be different as not all the time object will be changed, thus reference stays.

In short main idea was to reduce render() call, by reducing properties that are checked.

**Comment by zans-laksa** [comment]
Ok, now i get the idea.
Some props for most of the times are not changed at all and shouldn't be compared.
Also there are some props, that are marked as required, but are null or undefined during multiple render cycles, until the data is loaded.
But this solution is is the way to limit the count of render cycles and prevent some errors when incomplete data crashes components.
We just have to be sure that we correctly handle two component states: when the state is initial, props have default values and we display either loader or placeholder, and the state when the component has all data loaded and is fully rendered.
Thanks for the explanation. :)

**Comment by aleksandrakorolova** [comment]
I did ±5 tries on each page: 
Home page: the biggest result was 89 and the smaller - 52
Screenshot 2021-07-19 at 12 23 15

PLP: the biggest result was 69 and the smaller - 42
Screenshot 2021-07-19 at 12 20 29

PDP:  the biggest result was 80 and the smaller - 61
Screenshot 2021-07-19 at 11 29 02

Dashboard: the biggest result was 84 and the smaller - 61
Screenshot 2021-07-19 at 12 21 49

Address book: the biggest result was 84 and the smaller - 72
Screenshot 2021-07-19 at 11 45 07

Wishlist: the biggest result was 73 and the smaller - 55
Screenshot 2021-07-19 at 11 50 00

My orders: the biggest result was 83 and the smaller - 67
Screenshot 2021-07-19 at 11 46 23

My downloadable: the biggest result was 83 and the smaller - 64
Screenshot 2021-07-19 at 11 47 50

Cart: the biggest result was 65 and the smaller - 59
Screenshot 2021-07-19 at 11 54 45

Checkout: the biggest result was 79 and the smaller - 60
Screenshot 2021-07-19 at 11 55 27

**Comment by riha112** [comment]
Home page:
* First contentful paint - Ok
* Time to interact - OK
* Total block time - OK
* Speed Index - **Bad**:
**  **Reason Image sizes are to large** (not related to code performance) 
** Solution use lower resolution images, then score will be all good.
image

**Comment by riha112** [comment]
**PLP page:**
Problem:
* Speed index
Reason
* Page contains many DOM elements... we cannot really reduce them, but I have optimized as much as I can with shouldComponentUpdate to reduce render count
* Some attribute images doesn't have explicit width and height
image

**Comment by riha112** [comment]
PDP:
**Problem:**
* Speed index

**Reason:**
* Some attribute images doesn't have explicit width and height
image
